pkg/urbit/vere/saga.c

```diff
@@ -424,17 +424,83 @@ u3_saga_open(const c3_path* const pax_u, u3_meta* const met_u)
   }
 
   try_list(log_u->epo_u.lis_u = c3_list_init(), goto free_dir_entries);
-  u3_epoc *poc_u, *pre_u;
+  u3_epoc *poc_u;
   c3_w* lif_w = &met_u->lif_w;
+  c3_d  nex_d = 0;
+  c3_l  mex_l = 0;
   for ( size_t idx_i = 0; idx_i < ent_i; idx_i++ ) {
     c3_path_push(log_u->pax_u, ent_c[idx_i]);
     try_epoc(poc_u = u3_epoc_open(log_u->pax_u, lif_w), goto free_dir_entries);
     c3_path_pop(log_u->pax_u);
+
+    if ( NULL != san_i ) {
+      //  check epoch's event range for contiguousness with earlier epochs.
+      //  first epoch may start at arbitrary event nr, so we don't check it.
+      //
+      c3_d fir_d = u3_epoc_first_commit(poc_u);
+      if ( (0 != nex_d) && (nex_d != fir_d) ) {
+        u3l_log("sane: epoch snapshot not contiguous: "
+                "expected %" PRIu64 ", got %" PRIu64 ", in %s\n",
+                nex_d, fir_d, ent_c[idx_i]);
+        *san_i = 1;
+        goto next;
+      }
+      nex_d = u3_epoc_last_commit(poc_u) + 1;
+
+      //  load the snapshot so we can validate its mug
+      //
+      if ( 0 != mex_l ) {
+        //NOTE  careful, will clobber all existing noun references!
+        c3_path_push(log_u->pax_u, ent_c[idx_i]);
+        u3m_init();
+        u3e_load(c3_path_str(log_u->pax_u));
+        u3m_pave(c3n);
+        c3_path_pop(log_u->pax_u);
+
+        if ( mex_l != u3r_mug(u3A->roc) ) {
+          u3l_log("sane: epoch snapshot not contiguous, mug mismatch in %s\n",
+                  ent_c[idx_i]);
+          *san_i = 2;
+          goto next;
+        }
+      }
+
+      //  read the next mug from the last event in the epoch
+      //
+      if ( !u3_epoc_iter_open(poc_u, u3_epoc_last_commit(poc_u)) ) {
+        u3l_log("sane: failed to open epoch event log\n");
+        *san_i = 3;
+        goto next;
+      }
+      c3_y*  dat_y;
+      size_t dat_i;
+      if ( !u3_epoc_iter_step(poc_u, &dat_y, &dat_i) ) {
+        u3l_log("sane: failed to read last event\n");
+        *san_i = 4;
+        u3_epoc_iter_close(poc_u);
+        goto next;
+      }
+      u3_epoc_iter_close(poc_u);
+      if ( dat_i < 4 ) {
+        u3l_log("sane: strange event, too short\n");
+        *san_i = 5;
+        goto next;
+      }
+      mex_l = dat_y[0]
+            | dat_y[1] << 8
+            | dat_y[2] << 16
+            | dat_y[3] << 24;
+    }
+
+next:
     c3_list_pushb(log_u->epo_u.lis_u, poc_u, epo_siz_i);
     c3_free(poc_u);
     if ( lif_w ) {
       lif_w = NULL;
     }
+    if ( NULL != san_i && 0 != *san_i ) {
+      break;
+    }
```

Comment:
This should be `goto free_dir_entries`, not `break`, so that we fail if an epoch is not sane instead of continuing to open the event log.